### PR TITLE
feature: Downtime-less runtime updates

### PR DESCRIPTION
**Pull request by Yawning**
* [x] Document the design
 * [x] Implement the code
 * [x] Implement the tests

### CODE REVIEW

File: go/registry/api/runtime.go [hunk 1]
```diff
@@ -479,6 +482,102 @@ func (r *Runtime) ValidateBasic(strictVersion bool) error {
...
 	}
 
+	if len(r.Deployments) == 0 {
+		return fmt.Errorf("no deployment information specified")
+	}
+
+	return nil
+}
+
+func (r *Runtime) ActiveDeployment(now beacon.EpochTime) *VersionInfo {
```
Comment (kostko) [inline go/registry/api/runtime.go:492]: Add documentation comments.

File: go/runtime/registry/host.go [hunk 2]
```diff
@@ -85,6 +103,18 @@ func (n *RuntimeHostNode) WaitHostedRuntime(ctx context.Context) (host.RichRunti
 	return n.GetHostedRuntime(), nil
 }
 
+func (n *RuntimeHostNode) SetHostedRuntimeVersion(ctx context.Context, version version.Version) error {
```
Comment (kostko) [inline go/runtime/registry/host.go:106]: Add documentation comment.

File: go/common/node/node.go [hunk 3]
```diff
@@ -306,27 +306,45 @@ func (n *Node) IsExpired(epoch uint64) bool {
...
-// it. In case a runtime descriptor for the given runtime doesn't exist yet, a new one is created
-// appended to the list of supported runtimes and returned.
-func (n *Node) AddOrUpdateRuntime(id common.Namespace) *Runtime {
-	if rt := n.GetRuntime(id); rt != nil {
+// AddOrUpdateRuntime searches for an existing supported runtime descriptor
+// in Runtimes with the specified version and returns it. In case a
+// runtime descriptor for the given runtime and version doesn't exist yet,
+// a new one is created appended to the list of supported runtimes and
+// returned.
+func (n *Node) AddOrUpdateRuntime(id common.Namespace, version version.Version) *Runtime {
```
[the remark this answers is not in the document]
Comment (Yawning) [reply go/common/node/node.go:339]: When the runtime already exists, it returns the existing entry for update.  At least that was my rationale back in the day.

File: go/registry/api/api.go [hunk 4]
```diff
@@ -833,37 +858,82 @@ func verifyAddresses(params *ConsensusParameters, addressRequired bool, addresse
...
-			logger.Error("RegisterNode: trying to update runtimes, runtime Capabilities changed",
-				"current_runtime", curRtJSON,
-				"new_runtime", newRtJSON,
-			)
-			return false
+
+		// All versions present in currentMap for a runtime, that are
+		// also present in newMap, need to report identical capabilities.
+		//
+		// XXX: Should nodes be allowed to remove old versions?  Previous
```
[the remark this answers is not in the document]
Comment (Yawning) [reply go/registry/api/api.go:911]: No idea.

[the remark this answers is not in the document]
Comment (Yawning) [reply go/registry/api/api.go]: Fixed.

**Review by kostko — APPROVED** [review]

**Pull request merged**